Fix default/osdMessageQueue.cpp

anjohnson opened the pull request:
This branch contains my fixes for the message queue issue ([lp: #1868486](https://bugs.launchpad.net/epics-base/+bug/1868486)), but not the tests that I have also developed since they don't actually trigger any other bugs. Since I have already committed Heinz' fix for the issue that Mark Rivers was triggering I'm not as concerned whether this PR makes it into 3.15.8 or not, it would be nice to demonstrate the other issues that Michael thinks were present first, but I don't see that as essential.

anjohnson committed:
osdMessageQueue: Rename freeEventNode() -> destroyEventNode()

anjohnson committed:
osdMessageQueue: Return sooner on -ve timeout

It appears that previously a negative timeout actually implemented a
'wait forever', but the VxWorks and RTEMS implementations both check
for (timeout <= 0) and return immediately if nothing can be done
without waiting.

anjohnson committed:
osdMessageQueue: Clear eventNode before returning it

Introduced freeEventNode() which ensures eventNodes don't have a
signalled event in them before returning the node to the freeList.
Callers pass the status from epicsEventWaitWithTimeout() to indicate
whether it was signalled or not. If it timed out we must trigger it
and Wait to clear the event state.

anjohnson committed:
osdMessageQueue: This is the mirror of Heinz Junkes' earlier fix

When sending a message, if the queue is full so we have to wait, we
create a threadNode with an eventNode in it and stick it on the
sendQueue, then wait for a receiver to signal that event, waking us.
If we awoke due to a timeout but a receiver was actually waking us
up anyway (i.e. eventSent was set), we shouldn't give up.

anjohnson committed:
osdMessageQueue: Don't wake our sender until we're ready for it

Move the code that wakes up the next sending task to after we've
added our threadNode to the receiveQueue. He still has to wait for
us to release the Mutex though, so this might make no difference.
This commit also changes when we decrement the number of waiting
senders so it always happens immediately after a threadNode gets
taken off the sendQueue by the code that removed it.

anjohnson committed:
Cosmetic

mdavidsaver commented:
At a glance this looks like the right idea.  `freeEventNode()` should avoid returning a triggered epicsEvent to the free list, and its done safely under a lock.

(also to give Ben his due a9727fd5cbd6c27d4ecf6a61e6984537c3cbc5fe)

anjohnson committed:
osdMessageQueue: Undo change to -ve timeout handling

The internal mySend() and myReceive() routines do expect a timeout
of -1 to mean wait forever, see the epicsMessageQueueSend() and
epicsMessageQueueReceive() API routines.

anjohnson committed:
MessageQueue Tests: Extend Mark's sleep tests

anjohnson committed:
Allow/expect MinGW to fail epicsStackTraceTest #5

anjohnson committed:
eMQTest: Start each test with a new (empty) queue

If fastReceiver() took more than 0.01 seconds to exit,
sleepySender() might have pushed a second message onto
the queue after setting recvExit, so there would be an
extra message in the queue for the next test, which I
was seeing on Appveyor. That's my current theory...

anjohnson commented:
Yay, merging.
